hal/src/thumbv6m/sercom/v2/uart/reg.rs

```diff
@@ -0,0 +1,233 @@
+//! Register-level access to UART configuration
+
+use crate::sercom::v2::*;
+
+use super::{BaudMode, Parity, StopBits};
+use crate::time::Hertz;
+
+pub(super) trait RegisterInterface {
+    type Sercom: Sercom;
+    fn sercom(&self) -> &Self::Sercom;
```

Comment:
Unless you have some opposition, I would convert this to a type rather than a trait.

```rust
pub(super) type Registers<S: Sercom> {
    sercom: S,
}

impl<S: Sercom> Registers<S> {
    fn msb_first(&mut self, msb_first: bool) {
        // ...
    }
}
```
Then have the `Config` struct store a `Registers` struct *instead* of a `Sercom`. Basically, the `Registers` type acts to wrap the `Sercom` peripheral and expose a safer interface. You could even implement `Sync` for `Registers`, because you erase all of the interior mutability on the `Sercom` struct by requiring `&mut self` when appropriate.

This has an advantage, because you don't have any name clashes between public and private functions. It's also more consistent with what we're doing in the `clock::v2` module.

Note that I *had* to use a trait for `DynPin`, because `DynPin` erases the `PinId` type, so the corresponding `DynPinId` value can only be retrieved at run-time. A trait was the only way to implement it for both `Pin` and `DynPin`. But I still defined a `Registers<I: PinId>` type for `Pin`.

Reply:
By the way, I'm planning to implement this approach in `v2::spi` as well, when I get around to it.